src/codegen/function.rs

```diff
@@ -23,9 +25,92 @@ use std::{
+        || analysis.ret.nullable_return_is_error.is_some()
+    {
+        return None;
+    }
+
+    let return_type = if !analysis.outs.is_empty() {
+        let num_out_args = analysis
+            .outs
+            .iter()
+            .filter(|out| out.lib_par.array_length.is_none())
+            .count();
+        let num_out_sizes = analysis
+            .outs
+            .iter()
+            .filter(|out| out.lib_par.array_length.is_some())
+            .count();
+        // We need to differentiate between array(s)'s size arguments and normal ones. If we have 2
+        // "normal" arguments and one "size" argument, we still need to wrap them into "()" so we take
+        // that into account. If the opposite, it means that there are two arguments in any case so
+        // we need "()" too.
+        let num_outs = std::cmp::max(num_out_args, num_out_sizes);
+        match analysis.outs.mode {
+            Mode::Normal | Mode::Combined if num_outs == 1 => {
+                let array_lengths: Vec<_> = analysis
+                    .outs
+                    .iter()
+                    .filter_map(|out| out.lib_par.array_length)
+                    .collect();
+                let mut ret_param = None;
+
+                for out in analysis.outs.iter().filter(|out| !out.lib_par.is_error) {
+                    // The actual return value is inserted with an empty name at position 0
+                    if !out.lib_par.name.is_empty() {
+                        let mangled_par_name =
+                            crate::nameutil::mangle_keywords(out.lib_par.name.as_str());
+                        let param_pos = analysis
+                            .parameters
+                            .c_parameters
+                            .iter()
+                            .enumerate()
+                            .find_map(|(pos, orig_par)| {
+                                if orig_par.name == mangled_par_name {
+                                    Some(pos)
+                                } else {
+                                    None
+                                }
+                            })
+                            .unwrap();
+                        if array_lengths.contains(&(param_pos as u32)) {
+                            continue;
+                        }
+                    }
+                    ret_param = Some(out.lib_par.typ);
+                }
+                ret_param
+            }
+            _ => None,
+        }
+    } else {
+        analysis.ret.parameter.as_ref().map(|out| out.lib_par.typ)
+    };
+    if let Some(ref return_type) = return_type {
+        if let Some(parent_type_id) = parent_type_id {
+            if analysis.kind == library::FunctionKind::Method
+                && (!in_trait || only_declaration)
+                && parent_type_id == *return_type
+            {
+                return Some(format!("{}#[must_use]\n", comment_prefix));
+            }
+        }
+    }
```

Comment:
Completely unrelated to what you're working on, but this code could probably be re-used to check whether we should return Self or SomeTypeName; that should allow us to drop yet another `#[allow(some_clippy_warning)]`

Reply:
Ah true.